pkg/reader/reader.go

```diff
@@ -145,26 +160,60 @@ func (r *Reader) ParseStreamWithOptions(f io.ReadSeeker, o *Options) (*sbom.Docu
 
 	unserializer, err := GetFormatUnserializer(format)
 	if err != nil {
-		return nil, fmt.Errorf("getting format parser: %w", err)
+		return nil, fmt.Errorf("getting format parser for %s: %w", format, err)
 	}
 
 	// Build the listening chain of all the I/O sinks
-	var stream io.Reader = f
+	sinks := []io.Writer{}
 	for i := range o.Listeners {
-		stream = io.TeeReader(stream, o.Listeners[i])
+		sinks = append(sinks, o.Listeners[i])
 	}
 
+	// Create a byte counter to measure the size of the document
+	counter := byteCounter{
+		Count: new(int64),
+	}
+
+	// Create the hashers to tack the checksum of the original SBOM
+	hashers := map[sbom.HashAlgorithm]hash.Hash{
+		sbom.HashAlgorithm_SHA1:   sha1.New(), //nolint:gosec // SHA1 is required in SPDX2
+		sbom.HashAlgorithm_SHA256: sha256.New(),
+		sbom.HashAlgorithm_SHA512: sha512.New(),
+	}
+
+	if o.UnserializeOptions.TrackSource {
+		sinks = append(sinks, counter)
+		for i := range hashers {
+			sinks = append(sinks, hashers[i])
+		}
+	}
+
+	// We aggregate all the data sinks into a single multireader
+	// that gets a copy of all the bytes read from the stream.
+	multiwriter := io.MultiWriter(sinks...)
+	tee := io.TeeReader(f, multiwriter)
+
+	// Call the format unserializer
 	doc, err := unserializer.Unserialize(
-		stream, o.UnserializeOptions, r.Options.GetFormatOptions(unserializer),
+		tee, o.UnserializeOptions, r.Options.GetFormatOptions(unserializer),
 	)
 	if err != nil {
-		return nil, fmt.Errorf("unserializing: %w", err)
+		return nil, fmt.Errorf("unserializing %s: %w", format, err)
 	}
 
-	if doc.Metadata != nil {
-		doc.Metadata.SourceData, err = setSourceData(f, format)
-		if err != nil {
-			return nil, fmt.Errorf("setting origin info: %w", err)
+	// Protect in case the unserializer returns a nil document
+	if doc.Metadata == nil {
+		doc.Metadata = &sbom.Metadata{}
+	}
+
+	if o.UnserializeOptions.TrackSource {
+		doc.Metadata.SourceData = &sbom.SourceData{
+			Format: string(format),
+			Size:   *counter.Count,
+			Hashes: map[int32]string{},
+		}
+		for algo, hasher := range hashers {
+			doc.Metadata.SourceData.Hashes[int32(algo)] = fmt.Sprintf("%x", hasher.Sum(nil))
 		}
 	}
```

Comment:
Is this capturing the actual hash of the source data, or always computing the hash of `nil`? Based on my comment above, if using `*bytes.Buffer`:

```suggestion
	if o.UnserializeOptions.TrackSource {
		doc.Metadata.SourceData = &sbom.SourceData{
			Format: string(format),
			Size:   buf.Cap(),
			Hashes: map[int32]string{},
		}

		data := buf.Bytes()

		for algo, hasher := range hashers {
			doc.Metadata.SourceData.Hashes[int32(algo)] = fmt.Sprintf("%x", hasher.Sum(data))
		}
	}
```

Reply:
Ah no, we don't need to pass the data to the `Sum()` method because the hash.Hash interface implements io.Writer. the hashers are getting the SBOM through the `Write()` method as are the rest of the listeners.

Reply:
(The test has an example SBOM and their hashes if you want to verify)

Reply:
Oh ok, thanks for the info!